clause/label/label.go

```diff
@@ -33,7 +37,17 @@ func LabelProcessor(args map[string]interface{}) (elastic.Query, error) {
 		return nil, err
 	}
 
-	query := elastic.NewQueryStringQuery(realArgs.Label).Field("label")
+	if realArgs.Label == "" {
+		return nil, errors.New("No label was passed, cannot create clause.")
+	}
+
+	var processedQuery string
+	if realArgs.Exact {
+		processedQuery = realArgs.Label
+	} else {
+		processedQuery = clauseutils.AddImplicitWildcard(clauseutils.AddOrOperator(realArgs.Label))
+	}
+	query := elastic.NewQueryStringQuery(processedQuery).Field("label").QueryName(fmt.Sprintf("label: %q", realArgs.Label))
```

Comment:
This query name (elasticsearch's `_name`) is sort of experimental at this stage. Once we have a better idea how/if this will be used, it might be changed. What this could hopefully enable is passing the results back up to the UI annotated with which search clauses matched, which the UI can then use for highlighting or such. My thoughts here are that there's a couple options:

* Have the UI create some sort of opaque tokens which it passes down as arguments to the clause and expects to be added as names. This has the downside that the UI needs to keep track of those tokens, unless they're created in some sort of clever way that corresponds to display columns or something (the name can be any string).
* Have this create JSON which describes the clause that got passed in (sort of a more structured enhancement of the tiny bit I've added here). The things I'm not sure about with this are if it'll have issues with duplicate names, or if we need to have it pass in some sort of "context" (e.g. "this was the tenth clause in the 'any' list of a query which was the fourth element in a 'none' list of the top-level query") to be matched. All-in-all it seems like the first approach could be better, but this one, if it's possible, has a better chance of being basically automatic. So I'm not sure.

Reply:
Interesting! I hadn't known how kibana did it. From what I can see it looks like it could be reasonable, but I'd be curious what the UI side of the programming task thinks of the way the results come back as well. I'd need to adjust things a bit too, I think, to let clauses output something elsewhere than immediately in their hierarchy, but that's probably possible. I also wonder if we'd still want the named queries part of it -- really, they serve different but related needs if I'm understanding correctly, where the named queries will say which part(s) of the query matched but the highlighting will say which part(s) of the results matched.

Reply:
I agree that we should use Elasticsearch's highlighting for showing "which part(s) of the results matched".  I don't see any functional issues for the UI with the current structure coming from Kibana where the `highlight` portion is not immediately in the same hierarchy. However I would think the UI would have to check with each result if there is a `highlight` section and then replace the "plain" value with the highlight, which could be a performance hit with a lot of results and a lot of matching parts. I think ideally the UI would get the response back with the plain bits already replaced with their highlights, if that makes sense. 

As for "which part(s) of the query matched", I don't think I'm fully following.  Is there something in Elasticsearch that provides that information per match?  I didn't immediately see that, only what parts of the results matched.

Reply:
> I think ideally the UI would get the response back with the plain bits already replaced with their highlights, if that makes sense.

:+1: 

> As for "which part(s) of the query matched", I don't think I'm fully following.   

This feature sounds like _"icing on the cake"_.

Reply:
> I think ideally the UI would get the response back with the plain bits already replaced with their highlights, if that makes sense.

The service that'll eventually get built around this library can do that bit. What I meant as far as the "out of hierarchy" stuff is in the query building -- the highlighting stuff goes separately. I realize maybe it'd make more sense to have the UI request which fields should have highlighting, instead of having the search clauses do it, since the UI doesn't display every field, necessarily, so it'd be doing a lot of unnecessary highlighting for unshown fields. Does that make sense?

To be more precise, I imagine the UI making a request to the service with some parameter saying which fields should include highlighting information (say, `?highlight_fields=label,info-type,metadata` or something), then the service constructing the `highlight` clause of the query. Then, when returning results, the service would include something like `_highlighted_source` which is otherwise like the `_source` field but replacing fields with their highlighted versions. I'd assume it'd still return the `_source` and `highlight` fields as returned by ES.

I'm imagining that for these searches we'll pretty much always want to highlight the entire field by setting `number_of_fragments` to 0, otherwise some of the field might not appear in the highlighted section at all. This might be different if we do something for highlighting not-shown-by-default fields though, so it's worth mentioning I think.

> As for "which part(s) of the query matched", I don't think I'm fully following.

The part of ES which shows what part of the query matched the result is https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-named-queries-and-filters.html -- we add a `_name` to each thing corresponding to whatever we care about (presumably, to each clause), and then ES will give you back a `matched_queries` thing that tells you which of those queries (by name) were matched in each result. Initially I had figured this would be the way to do highlighting -- specify that certain clauses matched, then have the UI know which fields that means it should try to highlight. It's probably less useful with proper highlighting included instead.

Where I guess I could still see it being useful is if the UI wanted to add an extra column to show what matched, even if it wasn't in a field which would normally be shown. Then the UI would request highlighting for every field, and add names to all the queries that concern fields it doesn't normally show (for example, metadata). Then, when processing results, it would look at the matched_queries; if any are for a field it doesn't normally show, it looks up the appropriate field in the `highlights` and shows that in some way in the extra column.

Reply:
I like the idea of still enabling users to figure out what matched and why even if it's on something that's not typically shown as a column in the UI.  
In short, I think what you said in Slack is right and we should probably have a meeting to nail down the details 😄   Ramona or one of the other sci-analysts should definitely have input on this (though typically if you mention a use case where something could be useful, they agree we should have it).

Reply:
> I would think the UI would have to check with each result if there is a highlight section and then replace the "plain" value with the highlight, which could be a performance hit with a lot of results and a lot of matching parts.

It doesn't seem to me that it would be that big of a hit to let the client perform the highlighting. Even though a search can find 1000's of items, we use paging so that only 500 at a time are returned and processed by the UI.

I also don't have a problem with our search wrapper service doing that work for now, although if there's going to be a potential performance hit, having the service as the bottleneck for processing many search requests at once might be where we get hit; but then we can just launch more micro-services? 😁

Reply:
Yeah, I'm not sure how much it matters which place puts the highlights into their respective fields; that sort of mucking around with JSON probably isn't going to take that long wherever we put it. I think if we end up with a bottleneck, it's almost certainly going to be at whatever stage we actually calculate *what* to highlight, which in this case would be elasticsearch no matter what. It's also why I think we should only tell ES to do highlights for stuff where the calling service (the UI, at least initially) knows how to do something with it, and make sure we're doing the right thing as far as choosing fragment sizes and stuff like that. Which, once again, we probably should just meet to talk about I suppose. So, heh.

Reply:
That's a good point, Paul, I hadn't considered the paging.  👍